src/nested_json.rs

```diff
@@ -0,0 +1,397 @@
+use std::{fmt, mem};
+
+use anyhow::{anyhow, Result};
+use serde_json::map::Map;
+use serde_json::Value;
+
+use crate::utils::unescape;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum PathComponent {
+    Key(String, (usize, usize)),
+    Index(Option<usize>, (usize, usize)),
+}
+
+/// Parses a JSON path.
+///
+/// A JSON path is a string that satisfies the following syntax grammar:
+/// ```
+///   start: root_path path*
+///   root_path: (literal | index_path | append_path)
+///   literal: TEXT | NUMBER
+///
+///   path: key_path | index_path | append_path
+///   key_path: LEFT_BRACKET TEXT RIGHT_BRACKET
+///   index_path: LEFT_BRACKET NUMBER RIGHT_BRACKET
+///   append_path: LEFT_BRACKET RIGHT_BRACKET
+/// ```
+///
+/// Additionally, a backslash character can be used to:
+/// - Escape characters with especial meaning such as `\`, `[` and `]`.
+/// - Treat numbers as a key rather than as an index.
+pub fn parse_path(raw_json_path: &str) -> Result<Vec<PathComponent>> {
+    use PathComponent::*;
+    const SPECIAL_CHARS: &str = "=@:;[]\\";
+    let mut delims = vec![];
+    let mut backslashes = 0;
+
+    for (i, ch) in raw_json_path.char_indices() {
+        if ch == '\\' {
+            backslashes += 1;
+        } else {
+            if (ch == '[' || ch == ']') && backslashes % 2 == 0 {
+                delims.push((i, ch));
+            }
+            backslashes = 0;
+        }
+    }
+
+    if delims.is_empty() {
+        return Ok(vec![Key(
+            unescape(raw_json_path, SPECIAL_CHARS),
+            (0, raw_json_path.len()),
+        )]);
```

Comment:
This is a bit weird. Why would using a single quote vs a double quote lead to different results?
```
$ http --offline --print=B httpbin.org/json "\\4=6"
{
    "4": "6"
}
$ http --offline --print=B httpbin.org/json '\\4=6'
{
    "\\4": "6"
}
```

Reply:
That's a shell thing. With the double quotes the first backslash escapes the second so HTTPie sees `\4=6`. Single quotes don't handle a backslash specially at all, so it sees two backslashes, `\\4=6`. Compare:

```
$ echo '\\'
\\
$ echo "\\"
\
```

It's all about a single backslash, the double backslashes are because of shell and JSON. Another way to write it:
```
$ ~/p/xh/target/debug/xh --offline --print=B : '\0='
{
    "\\0": ""
}
$ http --offline --print=B : '\0='
{
    "0": ""
}
```

Reply:
It makes sense now. Thanks!